yjit/src/codegen.rs

```diff
@@ -4021,6 +4021,113 @@ fn jit_rb_false(
     true
 }
 
+
+/// Codegen for Kernel::is_a?
+fn jit_rb_kernel_is_a(
+    jit: &mut JITState,
+    ctx: &mut Context,
+    asm: &mut Assembler,
+    ocb: &mut OutlinedCb,
+    _ci: *const rb_callinfo,
+    _cme: *const rb_callable_method_entry_t,
+    _block: Option<IseqPtr>,
+    argc: i32,
+    known_recv_class: *const VALUE,
+) -> bool {
+    if argc != 1 {
+        return false;
+    }
+
+    // If this is a super class we might not know the class
+    if known_recv_class.is_null() {
+        return false;
+    }
+
+    // Important note: This works because we depend on known_recv_class.
+    // The left hand side here will always have the same type because there
+    // is a guard before this method call that ensures that is the case.
+    // We just guard that the right hand side does not change.
+
+    let sample_rhs = jit_peek_at_stack(jit, ctx, 0);
+    let sample_lhs = jit_peek_at_stack(jit, ctx, 1);
+
+    // If we are testing against a module, we can't be certain
+    // about the class hierarchy.
+    if !unsafe { RB_TYPE_P(sample_rhs, RUBY_T_CLASS) } {
+        return false;
+    }
+    let sample_is_a = unsafe { rb_obj_is_kind_of(sample_lhs, sample_rhs) == Qtrue };
+
+    let side_exit = get_side_exit(jit, ocb, ctx);
+    asm.comment("Kernel::is_a?");
+    asm.cmp(ctx.stack_opnd(0), sample_rhs.into());
+    asm.jne(counted_exit!(ocb, side_exit, send_is_a_class_mismatch));
```

Comment:
It might not be visible in railsbench, but it seems fairly common that the expected result for `is_a?` changes in a single callsite (e.g. [an example in rails](https://github.com/rails/rails/blob/6e721d79ba9458317c93bb02590bbaa25c5a3649/railties/lib/rails/application.rb#L223-L236) which doesn't seem to expect a stable result).

How about making this a chain guard?

Reply:
This uses the guard from send, so it's already a chain guard and guaranteed to be stable in this codegen method 🌈

Reply:
Ah interesting. Then we should keep the current code to avoid allocating an unnecessary `Branch` here. Thanks! 

It might be useful for readers like me to have a comment about it here.

```suggestion
    // The result should be stable because send uses jit_guard_known_klass.
    asm.jne(counted_exit!(ocb, side_exit, send_is_a_class_mismatch));
```

Reply:
> I have this big comment above bout this.

```
// Important note: This works because we depend on known_recv_class.
// The left hand side here will always have the same type because there
// is a guard before this method call that ensures that is the case.
// We just guard that the right hand side does not change.

```